staging/src/k8s.io/apiserver/pkg/server/hooks.go

```diff
@@ -176,6 +176,14 @@ func (s *GenericAPIServer) isPostStartHookRegistered(name string) bool {
 	return exists
 }
 
+// isPreShutdownHookRegistered checks whether a given PreShutdownHook is registered
+func (s *GenericAPIServer) isPreShutdownHookRegistered(name string) bool {
```

Comment:
don't like this very much. This makes the initialization more dynamic than it should be.

Reply:
How about move the audit PreShutdownHookRegister to pkg/master/master.go? So no need to use  `isPreShutdownHookRegistered` to check

Reply:
@sttts I found if we put addPreShutdownHook to genericapiserver.go, we must use this to prevent add hook die. Because many apiservers call `func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {`

Reply:
I think the following e2e case fail related to this.